File: HKBusETA/app/src/main/java/com/loohp/hkbuseta/utils/HTTPRequestUtils.java
/*
 * This file is part of HKBusETA.
 *
 * Copyright (C) 2023. LoohpJames <jamesloohp@gmail.com>
 * Copyright (C) 2023. Contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.loohp.hkbuseta.utils;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.nio.charset.StandardCharsets;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import javax.net.ssl.HttpsURLConnection;

public class HTTPRequestUtils {

    public static String getTextResponse(String link) {
        try {
            URL url = new URL(link);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setUseCaches(false);
            connection.setDefaultUseCaches(false);
            connection.addRequestProperty("User-Agent", "Mozilla/5.0");
            connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate");
            connection.addRequestProperty("Pragma", "no-cache");
            if (connection.getResponseCode() == HttpsURLConnection.HTTP_OK) {
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
                    return reader.lines().collect(Collectors.joining());
                }
            } else {
                return null;
            }
        } catch (IOException e) {
            return null;
        }
    }

    public static String getTextResponseWithPercentageCallback(String link, long customContentLength, Consumer<Float> percentageCallback) {
        return getTextResponseWithPercentageCallback(link, customContentLength, i -> i, percentageCallback);
    }

    public static String getTextResponseWithPercentageCallback(String link, long customContentLength, IOFunction<InputStream, InputStream> inputStreamTransform, Consumer<Float> percentageCallback) {
        try {
            URL url = new URL(link);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setUseCaches(false);
            connection.setDefaultUseCaches(false);
            connection.addRequestProperty("User-Agent", "Mozilla/5.0");
            connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate");
            connection.addRequestProperty("Pragma", "no-cache");
            if (connection.getResponseCode() == HttpsURLConnection.HTTP_OK) {
                float contentLength = customContentLength >= 0 ? customContentLength : connection.getContentLengthLong();
                try (InputStream inputStream = inputStreamTransform.apply(connection.getInputStream())) {
                    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
                    int readTotal = 0;
                    int nRead;
                    byte[] data = new byte[16384];
                    while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
                        readTotal += nRead;
                        buffer.write(data, 0, nRead);
                        percentageCallback.accept(Math.max(0F, Math.min(readTotal / contentLength, 1F)));
                    }
                    percentageCallback.accept(1F);
                    return new String(buffer.toByteArray(), 0, buffer.size(), StandardCharsets.UTF_8);
                }
            } else {
                return null;
            }
        } catch (IOException e) {
            return null;
        }
    }

    public static JSONObject getJSONResponse(String link) {
        return getJSONResponse(link, i -> i);
    }

    public static JSONObject getJSONResponse(String link, IOFunction<InputStream, InputStream> inputStreamTransform) {
        try {
            URL url = new URL(link);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setUseCaches(false);
            connection.setDefaultUseCaches(false);
            connection.addRequestProperty("User-Agent", "Mozilla/5.0");
            connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate");
            connection.addRequestProperty("Pragma", "no-cache");
            if (connection.getResponseCode() == HttpsURLConnection.HTTP_OK) {
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStreamTransform.apply(connection.getInputStream())))) {
                    String reply = reader.lines().collect(Collectors.joining());
                    return new JSONObject(reply);
                }
            } else {
                return null;
            }
        } catch (IOException | JSONException e) {
            return null;
        }
    }

    public static JSONObject postJSONResponse(String link, JSONObject body) {
        try {
            URL url = new URL(link);
            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setDoOutput(true);
            connection.setUseCaches(false);
            connection.setDefaultUseCaches(false);
            connection.addRequestProperty("User-Agent", "Mozilla/5.0");
            connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate");
            connection.addRequestProperty("Pragma", "no-cache");
            try (OutputStream os = connection.getOutputStream()) {
                byte[] input = body.toString().getBytes(StandardCharsets.UTF_8);
                os.write(input, 0, input.length);
            }
            if (connection.getResponseCode() == HttpsURLConnection.HTTP_OK) {
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()))) {
                    String reply = reader.lines().collect(Collectors.joining());
                    return new JSONObject(reply);
                }
            } else {
                return null;
            }
        } catch (IOException | JSONException e) {
            return null;
        }
    }

    public static InputStream getInputStream(String link) throws IOException {
        URLConnection connection = new URL(link).openConnection();
        connection.setConnectTimeout(20000);
        connection.setReadTimeout(20000);
        connection.setUseCaches(false);
        connection.setDefaultUseCaches(false);
        connection.addRequestProperty("User-Agent", "Mozilla/5.0");
        connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate");
        connection.addRequestProperty("Pragma", "no-cache");
        return connection.getInputStream();
    }

    public static byte[] download(String link) throws IOException {
        try (InputStream is = getInputStream(link)) {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            byte[] byteChunk = new byte[4096];
            int n;
            while ((n = is.read(byteChunk)) > 0) {
                baos.write(byteChunk, 0, n);
            }
            return baos.toByteArray();
        }
    }

    public static long getContentSize(String link) {
        try {
            URLConnection connection = new URL(link).openConnection();
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setUseCaches(false);
            connection.setDefaultUseCaches(false);
            connection.addRequestProperty("User-Agent", "Mozilla/5.0");
            connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate");
            connection.addRequestProperty("Pragma", "no-cache");
            if (connection instanceof HttpURLConnection) {
                ((HttpURLConnection) connection).setRequestMethod("HEAD");
            }
            return connection.getContentLengthLong();
        } catch (IOException e) {
            return -1;
        }
    }

    public static String getContentType(String link) {
        try {
            URLConnection connection = new URL(link).openConnection();
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setUseCaches(false);
            connection.setDefaultUseCaches(false);
            connection.addRequestProperty("User-Agent", "Mozilla/5.0");
            connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate");
            connection.addRequestProperty("Pragma", "no-cache");
            if (connection instanceof HttpURLConnection) {
                ((HttpURLConnection) connection).setRequestMethod("HEAD");
            }
            return connection.getContentType();
        } catch (IOException e) {
            return "";
        }
    }

}

File: HKBusETA/app/src/main/java/com/loohp/hkbuseta/utils/HTTPRequestUtils.kt
/*
 * This file is part of HKBusETA.
 *
 * Copyright (C) 2023. LoohpJames <jamesloohp@gmail.com>
 * Copyright (C) 2023. Contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package com.loohp.hkbuseta.utils

import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.stream.Collectors
import javax.net.ssl.HttpsURLConnection


fun getTextResponse(link: String?): String? {
    try {
        val url = URL(link)
        val connection = url.openConnection() as HttpsURLConnection
        connection.connectTimeout = 20000
        connection.readTimeout = 20000
        connection.useCaches = false
        connection.defaultUseCaches = false
        connection.addRequestProperty("User-Agent", "Mozilla/5.0")
        connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate")
        connection.addRequestProperty("Pragma", "no-cache")
        if (connection.responseCode == HttpsURLConnection.HTTP_OK) {
            BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
                return reader.lines().collect(Collectors.joining())
            }
        } else {
            return null
        }
    } catch (e: IOException) {
        return null
    }
}

fun getTextResponseWithPercentageCallback(link: String, customContentLength: Long, percentageCallback: (Float) -> Unit): String? {
    return getTextResponseWithPercentageCallback(link, customContentLength, { it }, percentageCallback)
}

fun getTextResponseWithPercentageCallback(link: String, customContentLength: Long, inputStreamTransform: (InputStream) -> InputStream, percentageCallback: (Float) -> Unit): String? {
    try {
        val url = URL(link)
        val connection = url.openConnection() as HttpsURLConnection
        connection.connectTimeout = 20000
        connection.readTimeout = 20000
        connection.useCaches = false
        connection.defaultUseCaches = false
        connection.addRequestProperty("User-Agent", "Mozilla/5.0")
        connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate")
        connection.addRequestProperty("Pragma", "no-cache")
        if (connection.responseCode == HttpsURLConnection.HTTP_OK) {
            val contentLength = (if (customContentLength >= 0) customContentLength else connection.contentLengthLong).toFloat()
            inputStreamTransform.invoke(connection.inputStream).use { inputStream ->
                val buffer = ByteArrayOutputStream()
                var readTotal = 0
                var nRead: Int
                val data = ByteArray(16384)
                while (inputStream.read(data, 0, data.size).also { nRead = it } != -1) {
                    readTotal += nRead
                    buffer.write(data, 0, nRead)
                    percentageCallback.invoke(0f.coerceAtLeast((readTotal / contentLength).coerceAtMost(1f)))
                }
                percentageCallback.invoke(1f)
                return String(buffer.toByteArray(), 0, buffer.size(), StandardCharsets.UTF_8)
            }
        } else {
            return null
        }
    } catch (e: IOException) {
        return null
    }
}

fun getJSONResponse(link: String): JSONObject? {
    return getJSONResponse(link) { it }
}

fun getJSONResponse(link: String, inputStreamTransform: (InputStream) -> InputStream): JSONObject? {
    try {
        val url = URL(link)
        val connection = url.openConnection() as HttpsURLConnection
        connection.connectTimeout = 20000
        connection.readTimeout = 20000
        connection.useCaches = false
        connection.defaultUseCaches = false
        connection.addRequestProperty("User-Agent", "Mozilla/5.0")
        connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate")
        connection.addRequestProperty("Pragma", "no-cache")
        if (connection.responseCode == HttpsURLConnection.HTTP_OK) {
            BufferedReader(InputStreamReader(inputStreamTransform.invoke(connection.inputStream))).use { reader ->
                val reply = reader.lines().collect(Collectors.joining())
                return JSONObject(reply)
            }
        } else {
            return null
        }
    } catch (e: IOException) {
        return null
    } catch (e: JSONException) {
        return null
    }
}

fun postJSONResponse(link: String, body: JSONObject): JSONObject? {
    try {
        val url = URL(link)
        val connection = url.openConnection() as HttpsURLConnection
        connection.connectTimeout = 20000
        connection.readTimeout = 20000
        connection.requestMethod = "POST"
        connection.setRequestProperty("Content-Type", "application/json")
        connection.doOutput = true
        connection.useCaches = false
        connection.defaultUseCaches = false
        connection.addRequestProperty("User-Agent", "Mozilla/5.0")
        connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate")
        connection.addRequestProperty("Pragma", "no-cache")
        connection.outputStream.use { os ->
            val input = body.toString().toByteArray(StandardCharsets.UTF_8)
            os.write(input, 0, input.size)
        }
        if (connection.responseCode == HttpsURLConnection.HTTP_OK) {
            BufferedReader(InputStreamReader(connection.inputStream)).use { reader ->
                val reply = reader.lines().collect(Collectors.joining())
                return JSONObject(reply)
            }
        } else {
            return null
        }
    } catch (e: IOException) {
        return null
    } catch (e: JSONException) {
        return null
    }
}

fun getInputStream(link: String): InputStream {
    val connection = URL(link).openConnection()
    connection.connectTimeout = 20000
    connection.readTimeout = 20000
    connection.useCaches = false
    connection.defaultUseCaches = false
    connection.addRequestProperty("User-Agent", "Mozilla/5.0")
    connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate")
    connection.addRequestProperty("Pragma", "no-cache")
    return connection.getInputStream()
}

fun download(link: String): ByteArray {
    getInputStream(link).use { `is` ->
        val baos = ByteArrayOutputStream()
        val byteChunk = ByteArray(4096)
        var n: Int
        while (`is`.read(byteChunk).also { n = it } > 0) {
            baos.write(byteChunk, 0, n)
        }
        return baos.toByteArray()
    }
}

fun getContentSize(link: String): Long {
    return try {
        val connection = URL(link).openConnection()
        connection.connectTimeout = 20000
        connection.readTimeout = 20000
        connection.useCaches = false
        connection.defaultUseCaches = false
        connection.addRequestProperty("User-Agent", "Mozilla/5.0")
        connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate")
        connection.addRequestProperty("Pragma", "no-cache")
        if (connection is HttpURLConnection) {
            connection.requestMethod = "HEAD"
        }
        connection.contentLengthLong
    } catch (e: IOException) {
        -1
    }
}

fun getContentType(link: String): String {
    return try {
        val connection = URL(link).openConnection()
        connection.connectTimeout = 20000
        connection.readTimeout = 20000
        connection.useCaches = false
        connection.defaultUseCaches = false
        connection.addRequestProperty("User-Agent", "Mozilla/5.0")
        connection.addRequestProperty("Cache-Control", "no-cache, no-store, must-revalidate")
        connection.addRequestProperty("Pragma", "no-cache")
        if (connection is HttpURLConnection) {
            connection.requestMethod = "HEAD"
        }
        connection.contentType
    } catch (e: IOException) {
        ""
    }
}
